### PR TITLE
AMP Observability Pattern

### PR DESCRIPTION
**Pull request by lewinkedrs**
### What does this PR do?

This PR provides an AMP observability pattern (module and example) to give customers insights into their AMP usage and spend.

Fixes issue #30 

### Motivation

Customers using the accelerator should have a module that provides insight and observability over the AMP deployment itself. This should not depend on any metrics within AMP itself.

### More

- [X] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [X] Yes, I have added a new example under [examples](https://github.com/aws-observability/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [ ] Yes, I have updated the [docs](https://github.com/aws-observability/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [X] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: examples/amp-monitoring/variables.tf [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  type        = string
+}
+
+variable "managed_grafana_workspace_id" {
+  description = "Amazon Managed Grafana (AMG) workspace ID"
+  type        = string
+  default     = ""
+}
+
+variable "enable_managed_grafana" {
```
Comment (bonclay7) [inline examples/amp-monitoring/variables.tf:24]: I think we should remove this for now. MG creation is not that straightforward. Or at least we should default it to false

File: examples/amp-monitoring/versions.tf [hunk 2]
```diff
@@ -0,0 +1,14 @@
+terraform {
+  required_version = ">= 1.1.0, < 1.3.0"
```
Comment (bonclay7) [inline examples/amp-monitoring/versions.tf:2]: `required_version = ">= 1.1.0"`

File: examples/amp-monitoring/main.tf [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  value       = grafana_folder.this.id
+}
+
+module "amp_monitor" {
+    source = "../../modules/workloads/amp-monitoring"
+    dashboards_folder_id = resource.grafana_folder.this.id
+    aws_region = local.region
+    managed_prometheus_workspace_id = var.managed_prometheus_workspace_id
+    depends_on = [
+    resource.grafana_folder.this
```
Comment (bonclay7) [inline examples/amp-monitoring/main.tf:59]: please run `terraform fmt` or run `pre-commit run -a` locally

File: examples/amp-monitoring/main.tf [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    amg_ws_endpoint = var.managed_grafana_workspace_id == "" ? "https://${module.managed_grafana[0].workspace_endpoint}" : "https://${data.aws_grafana_workspace.this[0].endpoint}"
+    amg_ws_id       = var.managed_grafana_workspace_id == "" ? split(".", module.managed_grafana[0].workspace_endpoint)[0] : var.managed_grafana_workspace_id
+    tags = {
+        Source = "github.com/aws-observability/terraform-aws-observability-accelerator"
+    }
+    name = "aws-observability-accelerator"
+}
+
+module "managed_grafana" {
+  count   = var.enable_managed_grafana ? 1 : 0
```
Comment (bonclay7) [inline examples/amp-monitoring/main.tf:31]: oh so you create a workspace and data source here. Why not extend the datasource config in the base module?
Examples should not have the main logic imo

File: examples/amp-monitoring/main.tf [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    managed_prometheus_workspace_id = var.managed_prometheus_workspace_id
+    depends_on = [
+    resource.grafana_folder.this
+    ]
+}
+
+module "billing" {
+  source = "../../modules/Billing"
+  providers = {
+    aws = aws.billing_region
```
Comment (bonclay7) [inline examples/amp-monitoring/main.tf:66]: nice! This is the concept i was talking to you about. we have to extend it to AMP and MG to support x-region

File: examples/amp-monitoring/README.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+1. Cloudwatch datasource on Grafana
+
+Open your Grafana workspace and under Configuration -> Data sources, you should see `aws-observability-accelerator-cloudwatch`. Open and click `Save & test`. You should see a notification confirming that the CloudWatch datasource is ready to be used on Grafana.
+
+2. Grafana dashboards
+
+Go to the Dashboards panel of your Grafana workspace. You should see a list of dashboards under the `AMP Monitoring Dashboards` folder.
+
+Open the `AMP Accelerator Dashboard` to see a visualization of the AMP workspace.
```
Comment (bonclay7) [inline examples/amp-monitoring/README.md:81]: If we move that to the base module we can reuse the same folder `Observability Accelerator Dashboards`

File: modules/billing/main.tf [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    return_data = "true"
+  }
+
+  metric_query {
+    id          = "m1"
+    return_data = "true"
+    metric {
+      metric_name = "Estimated Charges"
+      namespace   = "AWS/Billing"
+      period      = "21600"
```
Comment (bonclay7) [inline modules/billing/main.tf:22]: We should make variables with default values and allow customers to change them

File: modules/billing/main.tf [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      stat        = "Maximum"
+      unit        = "Count"
+
+      dimensions = {
+        ServiceName = "Prometheus"
+        Currencty = "USD"
```
Comment (bonclay7) [inline modules/billing/main.tf:28]: typo on "Currencty"

File: modules/billing/versions.tf [hunk 9]
```diff
@@ -0,0 +1,14 @@
+terraform {
+  required_version = ">= 1.1.0, < 1.3.0"
```
Comment (bonclay7) [inline modules/billing/versions.tf:2]: `required_version = ">= 1.1.0"`

File: modules/workloads/amp-monitoring/alarms.tf [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  threshold                 = var.active_series_threshold
+  alarm_description         = "This metric monitors AMP active series metrics"
+  insufficient_data_actions = []
+  metric_query {
+    id = "m1"
+    return_data = true
+    metric {
+      metric_name = "ResourceCount"
+      namespace   = "AWS/Usage"
+      period      = "120"
```
Comment (bonclay7) [inline modules/workloads/amp-monitoring/alarms.tf:15]: Let's define variables and default values

File: modules/workloads/amp-monitoring/alarms.tf [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  alarm_description         = "This metric monitors AMP ingestion rate"
+  insufficient_data_actions = []
+  metric_query {
+    id = "m1"
+    return_data = true
+
+    metric {
+      metric_name = "ResourceCount"
+      namespace   = "AWS/Usage"
+      period      = "120"
```
Comment (bonclay7) [inline modules/workloads/amp-monitoring/alarms.tf:44]: Let's define variables and default values

File: modules/workloads/amp-monitoring/versions.tf [hunk 12]
```diff
@@ -0,0 +1,14 @@
+terraform {
+  required_version = ">= 1.1.0, < 1.3.0"
```
Comment (bonclay7) [inline modules/workloads/amp-monitoring/versions.tf:2]: `required_version = ">= 1.1.0"`

**Review by bonclay7 — CHANGES_REQUESTED** [review]
Let's start the discussion with these and thanks for this!

**Comment by bonclay7** [comment]
@lewinkedrs can you check the pre-commit errors on github actions? There are few variables not necessary.

**Review by lewinkedrs — COMMENTED** [review]
Resolved 1 and 3 with recent commit. We will discuss how to implement 2 .

**Review by bonclay7 — APPROVED** [review]
Thanks @lewinkedrs! Feel free to squash and merge

File: modules/workloads/managed-prometheus-monitoring/main.tf [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+resource "grafana_data_source" "cloudwatch" {
+  type       = "cloudwatch"
+  name       = local.name
+  is_default = true
+  json_data {
+    default_region  = var.aws_region
+    sigv4_auth      = true
+    sigv4_auth_type = "workspace-iam-role"
```
Comment (jeromeinsf) [inline modules/workloads/managed-prometheus-monitoring/main.tf:18]: Are you sure this works?
According to https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#allowed_auth_providers the expected value should be `ec2_iam_role`

**Pull request merged**